### PR TITLE
refactor(core): Convert pipeline config to typescript (and remove d3 v3 in the process)

### PR DESCRIPTION
**Pull request by jrsquared**

### CODE REVIEW

**Comment by anotherchrisberry** [comment]
Nice. Only suggestion would be to convert the directive to a component - we only use directives now when we absolutely need to. Basically, you just convert the `link` to a controller, then use `$ctrl` in place of `scope`.

**Comment by jrsquared** [comment]
@anotherchrisberry Good call. Converted.

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+    return ['$scope', '$element', '$', '$window', 'pipelineGraphService', 'pipelineConfigValidator'];
+  }
+
+  public constructor(private $scope: any,
+                     private $element: JQuery,
+                     private $: any,
+                     private $window: ng.IWindowService,
+                     private pipelineGraphService: PipelineGraphService,
+                     private pipelineConfigValidator: PipelineConfigValidator
+                     ) {}
```
Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: formatting weirdness

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+  public onNodeClick: (node: IPipelineNode) => void;
+  public shouldValidate: boolean;
+
+  static get $inject() {
+    return ['$scope', '$element', '$', '$window', 'pipelineGraphService', 'pipelineConfigValidator'];
+  }
+
+  public constructor(private $scope: any,
+                     private $element: JQuery,
+                     private $: any,
```
Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: is this not also `jQuery`?

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    this.$scope.graphVerticalPadding = 11;
+    this.$scope.minExecutionGraphHeight = 40;
+    this.$scope.labelOffsetX = this.$scope.nodeRadius + 3;
+    this.$scope.labelOffsetY = this.$scope.nodeRadius + 10;
+    let graphId = this.pipeline ? this.pipeline.id : this.execution.id;
+
+    /**
+     * Used to draw inverse bezier curve between stages
+     */
+    const curvedLink = (d: any) => {
```
Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: Any reason to not make all these types of functions either `public` or `private` as necessary?  I mean, if we're going to Typescript, let's go all the way.

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts (hunk 2 above)
Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: Nope.

Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: Interesting.  TIL...

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts (hunk 3 above)
Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: I guess not? Pulled out all the internal functions.

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts (hunk 1 above)
Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: made 'less weird'

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts (hunk 2 above)
Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: Ah, you got me looking. It's a `JQueryStatic`. Good call.

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    pipeline: '=',
+    execution: '=',
+    viewState: '=',
+    onNodeClick: '=',
+    shouldValidate: '=',
+  };
+
+  public templateUrl: string = require('./pipelineGraph.component.html');
+  public controller: any = PipelineGraphController;
+  public controllerAs = '$ctrl';
```
Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: This is the default value for `controllerAs` on components.  No need to specify explicitly with the same value.

Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: 👍

File: app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+export class PipelineGraphController implements ng.IComponentController {
+  public pipeline: IPipeline;
+  public execution: IExecution;
+  public viewState: IExecutionViewState;
+  public onNodeClick: (node: IPipelineNode) => void;
+  public shouldValidate: boolean;
+  private pipelineValidations: IPipelineValidationResults = { pipeline: [], stages: [] };
+  private minLabelWidth = 100;
+
```
Comment (icfantv) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: extra lines of whitespace

Comment (jrsquared) [inline app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts]: removed

**Pull request merged**